In [1]:
#Import libraries
import pandas as pd
import numpy as np
import os

In [2]:
# Creating Path
path = r"C:\Users\David\Desktop\04-23 Instacart Basket Analysis"

In [3]:
# Import CSV file using path
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders.csv'), index_col = False)

In [4]:
# Importing products.csv file
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [5]:
# Importing department.csv file
df_dep = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'departments.csv'), index_col = False)

Step 2 of task

In [6]:
# Find another identifier variable in the df_ords dataframe that doesn’t need to be included in your analysis 
# as a numeric variable and change it to a suitable format.

df_ords['user_id'] = df_ords['user_id'].astype('str')

In [8]:
df_ords.head()

order_id user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329       1    prior             1          2                  8   
1   2398795       1    prior             2          3                  7   
2    473747       1    prior             3          3                 12   
3   2254736       1    prior             4          4                  7   
4    431534       1    prior             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

Step 3 from Task

In [9]:
# Look for a variable in your df_ords dataframe with an unintuitive name and change its name without overwriting the data frame.

df_ords.rename(columns = {'order_dow' : 'order_day_of_week'}, inplace = True)

In [10]:
df_ords.rename(columns = {'days_since_prior_order' : 'days_since_last_order'}, inplace = True)

In [11]:
df_ords.head()

order_id user_id eval_set  order_number  order_day_of_week  \
0   2539329       1    prior             1                  2   
1   2398795       1    prior             2                  3   
2    473747       1    prior             3                  3   
3   2254736       1    prior             4                  4   
4    431534       1    prior             5                  4   

   order_hour_of_day  days_since_last_order  
0                  8                    NaN  
1                  7                   15.0  
2                 12                   21.0  
3                  7                   29.0  
4                 15                   28.0

Step 4 from Task

In [12]:
df_ords['order_hour_of_day'].value_counts(dropna = False)

10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: order_hour_of_day, dtype: int64

# Hour 10 is the busiest hour of the day for placing orders with 288418 orders in that hour

Step 5 of Task

In [13]:
# Transpose the many columns to rows
df_dep.T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [14]:
df_dep_t = df_dep.T

In [15]:
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [16]:
new_header = df_dep_t.iloc[0]

In [17]:
new_header

0    department
Name: department_id, dtype: object

In [18]:
df_dep_t_new = df_dep_t[1:]

In [19]:
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [20]:
df_dep_t_new

0
1            frozen
2             other
3            bakery
4           produce
5           alcohol
6     international
7         beverages
8              pets
9   dry goods pasta
10             bulk
11    personal care
12     meat seafood
13           pantry
14        breakfast
15     canned goods
16       dairy eggs
17        household
18           babies
19           snacks
20             deli
21          missing

In [21]:
df_dep_t_new.columns = new_header

In [22]:
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [23]:
# Create a data dictionary
data_dict = df_dep_t_new.to_dict('index')

In [24]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [25]:
print(data_dict.get('4'))

{'department': 'produce'}


# Value of 4 in the department_id is produce

Task 6 of Task

In [26]:
# Create a subset containing only the required information.

df_breakfast = df_prods.loc[df_prods['department_id'] == 14]

In [27]:
df_breakfast

product_id                                       product_name  \
27             28                                  Wheat Chex Cereal   
33             34                                                NaN   
67             68                            Pancake Mix, Buttermilk   
89             90                                       Smorz Cereal   
210           211   Gluten Free Organic Cereal Coconut Maple Vanilla   
...           ...                                                ...   
49330       49326                            Cereal Variety Fun Pack   
49395       49391            Light and Fluffy Buttermilk Pancake Mix   
49547       49543                          Chocolate Cheerios Cereal   
49637       49633               Shake 'N Pour Buttermilk Pancake Mix   
49667       49663  Ultra Protein Power Crunch Peanut Butter N' Ho...   

       aisle_id  department_id  prices  
27          121             14    10.1  
33          121             14    12.2  
67          130             14    13.7  
89          121             14     3.9  
210         130             14     3.6  
...         ...            ...     ...  
49330       121             14     9.1  
49395       130             14     2.0  
49547       121             14    10.8  
49637       130             14    14.2  
49667        57             14    10.2  

[1116 rows x 5 columns]

Task 7 of Task

In [28]:
# find products from alcohol, deli, beverages, and meat/seafood
df_dinner = df_prods.loc[df_prods['department_id'].isin([5,7,12,20])]

In [29]:
df_dinner

product_id                                    product_name  aisle_id  \
2               3            Robust Golden Unsweetened Oolong Tea        94   
6               7                  Pure Coconut Water With Orange        98   
9              10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10             11                               Peach Mango Juice        31   
16             17                               Rendered Duck Fat        35   
...           ...                                             ...       ...   
49676       49672                          Cafe Mocha K-Cup Packs        26   
49679       49675             Cinnamon Dolce Keurig Brewed K Cups        26   
49680       49676                          Ultra Red Energy Drink        64   
49686       49682                              California Limeade        98   
49688       49684       Vodka, Triple Distilled, Twist of Vanilla       124   

       department_id  prices  
2                  7     4.5  
6                  7     4.4  
9                  7     8.4  
10                 7     2.8  
16                12    17.1  
...              ...     ...  
49676              7     6.5  
49679              7    14.0  
49680              7    14.5  
49686              7     4.3  
49688              5     5.3  

[7650 rows x 5 columns]

Task 8 of Task

# How many rows does the last dataframe you created have?

7650 rows by 5 columns

Task 9 of Task

In [30]:
# Extract all the information you can about this user_id 1
df_ords_user1 = df_ords[df_ords['user_id']=='1']

In [31]:
df_ords_user1

order_id user_id eval_set  order_number  order_day_of_week  \
0    2539329       1    prior             1                  2   
1    2398795       1    prior             2                  3   
2     473747       1    prior             3                  3   
3    2254736       1    prior             4                  4   
4     431534       1    prior             5                  4   
5    3367565       1    prior             6                  2   
6     550135       1    prior             7                  1   
7    3108588       1    prior             8                  1   
8    2295261       1    prior             9                  1   
9    2550362       1    prior            10                  4   
10   1187899       1    train            11                  4   

    order_hour_of_day  days_since_last_order  
0                   8                    NaN  
1                   7                   15.0  
2                  12                   21.0  
3                   7                   29.0  
4                  15                   28.0  
5                   7                   19.0  
6                   9                   20.0  
7                  14                   14.0  
8                  16                    0.0  
9                   8                   30.0  
10                  8                   14.0

Step 10 of Task

In [32]:
# What basic stats can you provide based on the information you have about this user?
df_ords_user1.describe()

order_id  order_number  order_day_of_week  order_hour_of_day  \
count  1.100000e+01     11.000000          11.000000          11.000000   
mean   1.923450e+06      6.000000           2.636364          10.090909   
std    1.071950e+06      3.316625           1.286291           3.477198   
min    4.315340e+05      1.000000           1.000000           7.000000   
25%    8.690170e+05      3.500000           1.500000           7.500000   
50%    2.295261e+06      6.000000           3.000000           8.000000   
75%    2.544846e+06      8.500000           4.000000          13.000000   
max    3.367565e+06     11.000000           4.000000          16.000000   

       days_since_last_order  
count              10.000000  
mean               19.000000  
std                 9.030811  
min                 0.000000  
25%                14.250000  
50%                19.500000  
75%                26.250000  
max                30.000000

Step 12 of Task

In [33]:
# Export wrangled orders data
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_wrangled.csv'))

Step 13 of Task

In [34]:
# Export wrangled department data
df_dep_t_new.to_csv(os.path.join(path, '02 Data','Prepared Data', 'departments_wrangled.csv'))